educe/rst_dt/learning/doc_vectorizer.py

```diff
@@ -217,28 +221,36 @@ def _extract_feature_vectors(self, doc):
         sf_cache = dict()
 
         for edu1, edu2 in edu_pairs:
+            feat_dict = dict()
             # retrieve info for each EDU
             edu_info1 = edu_info[edu1]
             edu_info2 = edu_info[edu2]
-
-            # extract and cache single features
-            try:
-                sf_cache1 = sf_cache[edu1]
-            except KeyError:
-                sf_cache[edu1] = list(sing_extract(edu_info1))
-                sf_cache1 = sf_cache[edu1]
-
-            try:
-                sf_cache2 = sf_cache[edu2]
-            except KeyError:
-                sf_cache[edu2] = list(sing_extract(edu_info2))
-                sf_cache2 = sf_cache[edu2]
-
-            # extract pair features
-            feats = list(pair_extract(edu_info1, edu_info2))
-            # re-emit single EDU features with a suffix
-            feats.extend(list(re_emit(sf_cache1, '_EDU1')))
-            feats.extend(list(re_emit(sf_cache2, '_EDU2')))
+            # gov EDU
+            if edu1 not in sf_cache:
+                sf_cache[edu1] = dict(sing_extract(edu_info1))
+            feat_dict['EDU1'] = dict(sf_cache[edu1])
+            # dep EDU
+            if edu2 not in sf_cache:
+                sf_cache[edu2] = dict(sing_extract(edu_info2))
+            feat_dict['EDU2'] = dict(sf_cache[edu2])
+            # pair
+            feat_dict['pair'] = dict(pair_extract(edu_info1, edu_info2))
+            # NEW
+            # product features
+            feat_dict['pair'].update(feat_prod(feat_dict['EDU1'],
+                                               feat_dict['EDU2'],
+                                               feat_dict['pair']))
+            # combine features
+            feat_dict['pair'].update(feat_comb(feat_dict['EDU1'],
+                                               feat_dict['EDU2'],
+                                               feat_dict['pair']))
+            # add suffix to single EDU features
+            feat_dict['EDU1'] = dict(re_emit(feat_dict['EDU1'].items(), '_EDU1'))
+            feat_dict['EDU2'] = dict(re_emit(feat_dict['EDU2'].items(), '_EDU2'))
+            # convert to list
```

Comment:
Oh right, that explains why my just-point-to-the-dict suggestion won't work.  You need a copy if you want to rename, right…